### PR TITLE
ci: run CI on main and release branches only

### PR DESCRIPTION
**Pull request by nsakaimbo**
### Description

Following migration of day-to-day PRs from Circle to TeamCity, this PR limits Circle usage to master/trunk branch merges and releases only.

### Blockers

Prior to landing this, we need to make sure a couple of things are in place:

- [x] TeamCity CI for all Calypso PRs (see #47478)
- ~~[ ] Automated Github reviews for TeamCity Linux jobs~~ (not necessary if the Desktop job is set as mandatory)

### CODE REVIEW

**Review by mokagio — APPROVED** [review]
Looks good. Comparing the checks this PR reports with the ones from #47441, I can see that the this one has no `wp-desktop-source` check and as a result none of the other platform specific builds, while the other does.

The same can be said looking at CircleCi itself

image


👍

File: .circleci/config.yml
Comment (nsakaimbo) [inline .circleci/config.yml]: Note this disables the linux build for merges to master. Building releases uses a different workflow and remains unchanged.

**Comment by scinos** [comment]
LGTM, but maybe we should run TC and CircleCI in parallel for a few days until we are sure the TC build is working correctly?

**Review by scinos — APPROVED** [review]

**Comment by nsakaimbo** [comment]
> maybe we should run TC and CircleCI in parallel for a few days until we are sure the TC build is working correctly?

Good idea. I'll hold off on merging this for a few days so we can see how things go.

**Comment by mokagio** [comment]
@nsakaimbo the title says "on main branch and releases only", but the branches in the filter are only `master` and `trunk`.

A thought just came to mind, would it be worth adding `release/*` as well? 

The reason I'm suggesting this is that since the deployment happens on CircleCI when we push a tag, it would be good to make sure the build for that commit already works on CircleCI, in order to avoid a time consuming "fix attempt, push new tag, see if it works..." workflow if the tag build fails for some reason. Of course, if it builds on Team City it's unlikely it'll fail on CircleCI. Still, there are quite a few moving parts in the CI setup.

We do few builds on the release branches, so the extra load shouldn't negligible (other than for the release manager how has to wait for the build on the release branch to pass before pushing the tag, then wait for the tag to build, too, but that's already how we work so no regression there).

**Comment by nsakaimbo** [comment]
> would it be worth adding release/* as well?

Thanks for catching this, @mokagio! Agree it might be helpful to permit release branch jobs, as having to re-tag things would be a pain. I went ahead and added release branch matching in f80f607db55467c9fac864cfa5f38d63058ded8e.

**Comment by mokagio** [comment]
Thanks @nsakaimbo!

**Comment by nsakaimbo** [comment]
Update: seems like the Desktop TC job is working as expected. However I'm holding off on landing this until a bug with the Calypso Circle webhooks is fixed. Context: main branch was recently migrated from `master` to `trunk`, and since then Calypso Circle CI has not been kicked off in the `trunk` branch. 

We narrowed down the problem to a missing step in the migration, which is described in [this Circle blog post](https://support.circleci.com/hc/en-us/articles/360040969693-VCS-Default-Branch-Isn-t-Reflected-on-CircleCI). I believe folks from Team Calypso are looking into this and hopefully master branch jobs start getting triggered again soon, at which point we can get this into `trunk`.

**Pull request merged**
